File: src/main/java/edu/psu/ist/plato/kaiming/Loop.java
package edu.psu.ist.plato.kaiming;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import edu.psu.ist.plato.kaiming.util.SetUtil;
import edu.psu.ist.plato.kaiming.util.Tuple;

public class Loop<T extends Entry> {
    private BasicBlock<T> mHeader;
    private Set<BasicBlock<T>> mBody;
    
    private Loop(BasicBlock<T> header, Set<BasicBlock<T>> loop) {
        mHeader = header;
        mBody = loop;
    }
    
    public static <T extends Entry> List<Loop<T>> detectLoops(CFG<T> cfg) {
        List<Loop<T>> ret = new ArrayList<>(); 
        
        Map<BasicBlock<T>, Set<BasicBlock<T>>> dominators = new HashMap<>();
        Set<BasicBlock<T>> tmp = new HashSet<>(), all;
        cfg.iterator().forEachRemaining(x -> tmp.add(x));
        all = Collections.unmodifiableSet(tmp);
        all.forEach(x -> dominators.put(x, all));
        List<Set<BasicBlock<T>>> singletons = all.stream().
                map(x -> Collections.unmodifiableSet(new HashSet<>(Arrays.asList(x)))).
                collect(Collectors.toList());
        boolean dirty = true;
        while (dirty) {
            dirty = false;
            for (Set<BasicBlock<T>> singleton : singletons) {
                BasicBlock<T> bb = singleton.iterator().next();
                Set<BasicBlock<T>> n =
                        bb.predecessors().stream().map(x -> dominators.get(x)).
                        reduce(SetUtil::intersection).orElse(singleton);
                
                if (!n.contains(bb))
                    n = SetUtil.add(n, bb);
                
                if (!n.equals(dominators.get(bb))) {
                    dirty = true;
                    dominators.put(bb, new HashSet<>(n));
                }
            }
        }
        List<Tuple<BasicBlock<T>, BasicBlock<T>>> backEdges = new ArrayList<>();
        for (Map.Entry<BasicBlock<T>, Set<BasicBlock<T>>> 
            entry : dominators.entrySet()) {
            BasicBlock<T> key = entry.getKey();
            Set<BasicBlock<T>> value = entry.getValue();
            backEdges.addAll(
                    SetUtil.intersection(new HashSet<>(key.successors()), value).
                    stream().map(x -> new Tuple<>(key, x)).collect(Collectors.toList()));
        }
        for (Tuple<BasicBlock<T>, BasicBlock<T>> t : backEdges) {
            ret.add(findLoopNodes(t, 
                    all.stream().filter(x -> dominators.get(x).contains(t.second)).
                    collect(Collectors.toSet())));
        }
        return ret;
    }
    
    private static <T extends Entry> Loop<T> findLoopNodes(
            Tuple<BasicBlock<T>, BasicBlock<T>> backEdge,
            Set<BasicBlock<T>> candidates) {
        Set<BasicBlock<T>> ret = new HashSet<>();
        ret.add(backEdge.first);
        ret.add(backEdge.second);
        for (BasicBlock<T> b : candidates) {
            if (ret.contains(b))
                continue;
            reachable(b, backEdge.first, candidates, ret, new HashSet<>());
        }
        return new Loop<T>(backEdge.second, ret);
    }
    
    private static <T extends Entry> boolean reachable(BasicBlock<T> start,
            BasicBlock<T> end, Set<BasicBlock<T>> candidates,
            Set<BasicBlock<T>> ret, Set<BasicBlock<T>> visited) {
        boolean canReach;
        if (start.equals(end)) {
            canReach = true;
        } else {
            canReach = false;
            visited.add(start);
            for (BasicBlock<T> s : start.successors()) {
                if (!candidates.contains(s))
                    continue;
                if (visited.contains(s)) {
                    canReach = canReach || ret.contains(s);
                    continue;
                }
                if (reachable(s, end, candidates, ret, visited)) {
                    canReach = true;
                }
            }
        }
        if (canReach)
            ret.add(start);
        return canReach;
    }
    
    public BasicBlock<T> header() {
        return mHeader;
    }
    
    public Set<BasicBlock<T>> loop() {
        return mBody;
    }
    
    @Override
    public String toString() {
        StringBuilder b = new StringBuilder();
        b.append("(");
        b.append(mHeader.label());
        b.append(" : [ ");
        mBody.forEach(x -> { b.append(x.label()); b.append(" "); });
        b.append("])");
        return b.toString();
    }
}

File: src/main/scala/edu/psu/ist/plato/kaiming/Loop.scala
package edu.psu.ist.plato.kaiming

case class Loop[A <: Arch] private (header: BBlock[A], body: Set[BBlock[A]]) {
  
  override def toString = {
    val b = new StringBuilder()
    b.append("(");
    b.append(header.label);
    b.append(": [ ");
    body.foreach {x => { b.append(x.label); b.append(" "); }}
    b.append("])");
    b.toString();
  }
  
}

object Loop {
  
  def detectLoops[T <: Arch](cfg: CFG[T]): List[Loop[T]] = {
    val allBBs = cfg.blocks.toSet
    val initDominators = allBBs.foldLeft(Map[BBlock[T], Set[BBlock[T]]]()) {
      (map, bb) => map + (bb -> allBBs)
    }
    val singletons = allBBs.map { x => (x, Set(x)) }
    def computeDoms(input: (Boolean, Map[BBlock[T], Set[BBlock[T]]]))
     : Map[BBlock[T], Set[BBlock[T]]] = input match {
      case (stop, in) =>
      if (stop)
        in
      else
        computeDoms(singletons.foldLeft((true, in)) {
          case ((stop, map), (bb, singleton)) => {
            val n = cfg.predecessors(bb).map(x => in.get(x)).flatten
            val newDomSet = if (n.size == 0) singleton else n.reduce(_&_) + bb
            (stop && newDomSet.equals(in.get(bb).orNull), in + (bb -> newDomSet))
          }
        })
    }
    val newDominators = computeDoms((false, initDominators))
    val backEdges = newDominators.foldLeft(List[(BBlock[T], BBlock[T])]()) {
      case (l, (k, v)) => l ++ (cfg.successors(k) & v).map { x => (k, x) }
    }
    backEdges.map { 
      x => 
        findLoopNodes(cfg, x, 
            allBBs.filter {
              y => newDominators.get(y).orNull.contains(x._2)
            })}
  }
  
  private def findLoopNodes[T <: Arch](cfg: CFG[T],
      backEdge: (BBlock[T], BBlock[T]),
      candidates: Set[BBlock[T]]): Loop[T] = {
    new Loop(
        backEdge._2,
        candidates.foldLeft(Set(backEdge._1, backEdge._2)) {
          (set, bb) =>
            if (set.contains(bb))
              set
            else
              reachable(cfg, bb, backEdge._1, candidates, set, Set[BBlock[T]]())._2 
    })
  }
  
  private def reachable[T <: Arch](cfg: CFG[T], start: BBlock[T], end: BBlock[T],
      candidates: Set[BBlock[T]], ret: Set[BBlock[T]],
      visited: Set[BBlock[T]]): (Boolean, Set[BBlock[T]]) = {
    if (start.equals(end))
      (true, ret + start)
    else {
      cfg.successors(start).foldLeft((false, ret)) {
        case ((canReach, set), bb) =>
          if (!candidates.contains(bb))
            (canReach, set)
          else if (visited.contains(bb))
            (canReach || ret.contains(bb), set)
          else 
            reachable(cfg, bb, end, candidates, ret, visited + start)
      }
    }
  }
  
}